### ISSUE TITLE
switch to json5

### ISSUE THREAD

**Issue by Laserlicht** [vcmi/vcmi#3604]
**Describe your proposal**
Just for discussion: Should we switch our config to json5?

The problem is that we use additional "features" in our json that not supported by the JSON standard. This causes problem in every case, where real json, standard confirm files are expected. For example editors and IDEs which therefore rightly mark the json document as incorrect and show errors.

Basically our used features are supported by json5 and our json files can be validated as json5. Basically they are already json5 but with the incorrect file ending. :)

There are already some big players using json5. E.g. Chromium: https://github.com/chromium/chromium/blob/feb3c9f670515edf9a88f185301cbd7794ee3e52/third_party/blink/renderer/platform/runtime_enabled_features.json5

What do you think?

**Additional context**
We should still able to load `.json`-Files for backward compatibility, but prefer `.json5`-File. And rename jsons in main repo as well as for new / reworked mods.

**Comment by IvanSavenko** [comment]
Might be a good idea, but not a quick one. I think some IDE's (e.g. VS code) have 'jsonc" format (Json with comments) which matches our json parser.

And from what I see, json5 is *way* more than json with comments. Here is their example:
```json5
{
  // comments
  unquoted: 'and you can quote me on that',
  singleQuotes: 'I can use "double quotes" here',
  lineBreaks: "Look, Mom! \
No \\n's!",
  hexadecimal: 0xdecaf,
  leadingDecimalPoint: .8675309, andTrailing: 8675309.,
  positiveSign: +1,
  trailingComma: 'in objects', andIn: ['arrays',],
  "backwardsCompatible": "with JSON",
}
```
Frankly this is not extension of json format but a completely new format.

So while our json files can be viewed as valid json5 format, json5 validator can NOT be used to validate our files. Because if people/modders will try to use any json5 feature other than comments json5 will okay this file but our json parser would fail.

**Comment by Laserlicht** [comment]
> Frankly this is not extension of json format but a completely new format.

Yes, just created a PR to test such cases. :)

> So while our json files can be viewed as valid json5 format, json5 validator can NOT be used to validate our files. Because if people/modders will try to use any json5 feature other than comments json5 will okay this file but our json parser would fail.

Yes, of course our JSON parser supports not every feature. So either a non-standard json parser, or an incomplete json5 parser. :)

By the way: Apple seems to have native JSON5 support: https://developer.apple.com/documentation/foundation/jsondecoder/3766916-allowsjson5

**Comment by IvanSavenko** [comment]
> Yes, just created a PR to test such cases. :)

Sorry, which one?
- #3605: Validates our json's as json5, BUT will okay all files that use json5 features that we don't support. Our current validator matches our parser way better.
- #3606: Renames references to json to json5. Potentially misleading since people might use json5 features that we don't support.

**Comment by Laserlicht** [comment]
https://github.com/vcmi/vcmi/pull/3606

:)

**Comment by IvanSavenko** [comment]
> #3606

Renames references to json to json5. Potentially misleading since people might use json5 features that we don't support.

Or are you planning to extend our json parser to support json5?

**Comment by IvanSavenko** [comment]
And there is also our amazing Android launcher which only accepts strict Json format. Hopefully it will finally die in 1.5, but that is yet to happen.

**Comment by Laserlicht** [comment]
> > #3606
> 
> Renames references to json to json5. Potentially misleading since people might use json5 features that we don't support.

Yes, your right. We should support (or at least warn) the additional features first...

**Comment by Alexander-Wilms** [comment]
Then maybe rename the files to `.jsonc` ? At least for VSCode that stops comments from being treated like errors.

**Comment by IvanSavenko** [comment]
Actually it looks like we can't.

Some players install vcmi by simply overwriting old version (mostly on Windows). And when VCMI looks for a file 'config/xxx.json' it actually looks for *any* such file. And in case of overwrite, player would have both old json file and new jsonc file.

**Comment by IvanSavenko** [comment]
Mostly implemented in PR #3618
Some bits are still missing, for example - multi-line comments - ` /* ... */ `

However:
- game still uses only `.json` extension - to prevent issues from installing game or mod on top of older version, leading to both `.json` and `.json5` files present in install directory.
- all files in config directory same format as before - json with comments.

**Comment by thiscris** [comment]
> and when VCMI looks for a file 'config/xxx.json' it actually looks for _any_ such file.
Wouldn't the change to .json5 mean that VCMI would be looking for .json5 files so even if somebody makes the mistake to update by overwriting, the old .json files would be ignored?

**Comment by IvanSavenko** [comment]
No, they won't be.
I actually remember some mod having similar issue: modder replaced bmp icons with png icons without deleting bmp's. However from vcmi point of view bmp files have same priority as png's, so which icon will be used is pretty much random.

Same issue will be if we were to accept `.json5` extension in addition to `.json`.
For example, if modder adds file `myMod/creatures/newCreature.json5` then even if he specifies filename as `.json5` in his mod.json, vcmi will still consider both `.json` and `.json5` when searching for the file.

So for now VCMI accepts json5 format, but still requires json extension. If you want to use json5 syntax highlight - configure your IDE to use json5 for vcmi files
